arch/aarch64/crc32_acle.c

```diff
@@ -15,66 +15,80 @@
 #ifdef __linux__
 #  include <stddef.h>
 #endif
+#include "zutil.h"
 
 uint32_t crc32_acle(uint32_t crc, const unsigned char *buf, uint64_t len) {
-    register uint32_t c;
-    register const uint16_t *buf2;
-    register const uint32_t *buf4;
-    register const uint64_t *buf8;
+    uint32_t c = ~crc;
 
-    c = ~crc;
     if (len && ((ptrdiff_t)buf & 1)) {
         c = __crc32b(c, *buf++);
         len--;
     }
 
-    if ((len > 2) && ((ptrdiff_t)buf & 2)) {
-        buf2 = (const uint16_t *) buf;
-        c = __crc32h(c, *buf2++);
-        len -= 2;
-        buf4 = (const uint32_t *) buf2;
-    } else {
-        buf4 = (const uint32_t *) buf;
+    if ((len > sizeof(uint16_t)) && ((ptrdiff_t)buf & 2)) {
+        uint16_t val;
+        MEMCPY(&val, buf, sizeof(val));
+        c = __crc32h(c, val);
+        buf += sizeof(val);
+        len -= sizeof(val);
     }
 
-    if ((len > 4) && ((ptrdiff_t)buf & 4)) {
```

Comment:
I think this should have ```buf4 & 4``` at the end instead... See line 37 of original code.

Reply:
Both `buf & 4` and `buf4 & 4` are equivalent:
on line 37 of original code `buf4 = buf`,
on line 35 of original code `buf4 = buf2` and on lines 32 and 33 we have `buf2 = buf + sizeof(uint16_t)`.
The bit we are looking at with `buf & 4` or with `buf4 & 4` is the same.

Reply:
I agree that it is less confusing to read `buf4 & 4`.  I will prepare a separate patch.